contrib/clients/xloadimage/xloadimage.man
.TH XLOADIMAGE 1 "7 October 1989"
.SH NAME
xloadimage, xsetbg, xview \- load images into an X11 window or onto
the root window
.SH SYNOPSIS
\fIxloadimage\fR [global_options] {[image_options] image ...}
.SH DESCRIPTION
\fIXloadimage\fR displays images in an X11 window or loads them onto
the root window.  See the \fIIMAGE TYPES\fR section below for
supported image types.  Compressed images will automatically be
uncompressed.
.PP
If the destination display cannot support the number of colors in the
image, the image will be dithered (monochrome destination) or have its
colormap reduced (color destination) as appropriate.  This can also be
done forcibly with the \fI-dither\fR and \fI-colors\fR options.
.PP
If more than one image is to be loaded, they will be merged into a
single image.  The \fI-at\fR and \fI-center\fR options control where
subsequent images will be loaded onto the initial image.  Any
combination of color image depths and/or monochrome images may be
loaded at one time.
.PP
A variety of image manipulations can be specified, including
brightening, clipping, dithering, depth-reduction, and zooming.  Most
of these manipulations have simple implementations; speed was opted
for above accuracy.
.PP
If you are viewing a large image in a window, the initial window will
be at most 90% of the size of the display.  You may move the image
around in the window by dragging with the first mouse button.  The
cursor will indicate which directions you may drag, if any.  You may
exit the window by typing 'q' or 'Q' when the keyboard focus is on the
window.
.PP
\fIXsetbg\fR is equivalent to \fIxloadimage -onroot -quiet\fR and
\fIxview\fR is equivalent to \fIxloadimage -view -verbose\fR.
.SH GLOBAL OPTIONS
The following options affect the global operation of \fIxloadimage\fR.
They may be specified anywhere on the command line.
.TP 8
-border \fIcolor\fR
This sets the background portion of the window which is not covered by
any images to be \fIcolor\fR.
.TP
-colors \fIn\fR
Specify the maximum number of colors to use in the image.  This is a
way to forcibly reduce the depth of an image.
.TP
-display \fIdisplay_name\fR
X11 display name to send the image(s) to.
.TP
-geometry \fIWxH[{+-X}{+-}Y]\fR
This sets the size of the window onto which the images are loaded to a
different value than the size of the image.  When viewing an image in
a window, this can be used to reduce the size of the destination
window.  When loading an image onto the root window, this option
controls the size of the pixmap which will be loaded onto the root.
If the size is smaller than that of the display, the image will be
replicated.
.TP
-install
Forcibly install the image's colormap when the window is focused.
This violates ICCCM standards and only exists to allow operation with
naive window managers.  Use this option only if your window manager
does not install colormaps properly.
.TP
-list
List the images which are along the image path.
.TP
-help
Displays a short summary of xloadimage command line syntax.
.TP
-onroot
Load image(s) onto the root window instead of viewing in a window.
This is the opposite of \fI-view\fR.  \fIXSetbg\fR has this option set
by default.
.TP
-path
Displays the image path and image suffixes which will be used when
looking for images.  These are loaded from ~/.xloadimagerc and
optionally from a systemwide file (normally /usr/lib/xloadimagerc).
.TP
-quiet
Forces \fIxloadimage\fR and \fIxview\fR to be quiet.  This is the
default for \fIxsetbg\fR, but the others like to whistle. 
.TP
-supported
List the supported image types. 
.TP
-verbose
Causes \fIxloadimage\fR to be talkative, telling you what kind of
image it's playing with and any special processing that it has to do. 
This is the default for \fIxview\fR and \fIxloadimage\fR. 
.TP
-view
View image(s) in a window.  This is the opposite of \fI-onroot\fR and
the default for \fIxsetbg\fR. 
.SH IMAGE OPTIONS
The following options may preceed each image.  These options are
local to the image they preceed. 
.TP
-background \fIcolor\fR
Use \fIcolor\fR as the background color instead of the default
(usually white but this depends on the image type) if you are
transferring a monochrome image to a color display. 
.TP
-brighten \fIpercentage\fR
Specify a percentage multiplier for a color image's colormap.  A value
of more than 100 will brighten an image, one of less than 100 will
darken it. 
.TP
-center
Center the image on the first image loaded.  If this is an option to
the first image, and the \fI-onroot\fR option is specified, the image
will be centered on the display background. 
.TP
-at \fIX\fR,\fIY\fR
Indicates coordinates to load the image at on the first image.  If
this is an option to the first image, and the \fI-onroot\fR option is
specified, the image will be loaded at the given location on the
display background. 
.TP
-clip \fIX\fR,\fIY\fR,\fIW\fR,\fIH\fR
Clip the image before loading it.  \fIX\fR and \fIY\fR define the
upper-left corner of the clip area, and \fIW\fR and \fIH\fR define the
extents of the area.  A zero value for \fIW\fR or \fIH\fR will be
interpreted as the remainder of the image. 
.TP
-dither
Force halftone dithering of a color image when displaying on a color
display.  This happens by default when viewing color images on a
monochrome display.  This option is ignored on monochrome images. 
.TP
-foreground \fIcolor\fR
Use \fIcolor\fR as the foreground color instead of black if you are
transferring a monochrome image to a color display.  This can also be
used to invert the foreground and background colors of a monochrome
image. 
.TP
-name \fIimage_name\fR
Force the next argument to be treated as an image name.  This is
useful if the name of the image is \fI-dither\fR, for instance. 
.TP
-xzoom \fIpercentage\fR
Zoom the X axis of an image by \fIpercentage\fR.  A number greater
than 100 will expand the image, one smaller will compress it.  A zero
value will be ignored. 
.TP
-yzoom \fIpercentage\fR
Zoom the Y axis of an image by \fIpercentage\fR.  See \fI-xzoom\fR for
more information. 
.TP
-zoom \fIpercentage\fR
Zoom both the X and Y axes by \fIpercentage\fR.  See \fI-xzoom\fR for
more information. 
.SH EXAMPLES
To load the rasterfile "my.image" onto the background and replicate
it to fill the entire background:
.sp
.ti +5
xloadimage my.image
.PP
To load a monochrome image "my.image" onto the background, using red
as the foreground color, replicate the image, and overlay
"another.image" onto it at coordinate (10,10):
.sp
.ti +5
xloadimage -foreground red my.image -at 10,10 another.image
.PP
To center the rectangular region from 10 to 110 along the X axis and
from 10 to the height of the image along the Y axis:
.sp
.ti +5
xloadimage -center -clip 10,10,100,0 my.image
.SH PATHS AND EXTENSIONS
The file ~/.xloadimagerc (and optionally a system-wide file which is
system-specific) defines the path and default extensions that
.I xloadimage
will use when looking for images.  The file can have two statements:
"path=" and "extension=" (the equals signs must follow the word with
no spaces between).  Everything following the "path=" keyword
will be prepended to the supplied image name if the supplied name does
not specify an existing file.  The paths will be searched in the order
they are specified.  Everything following the "extension=" keyword
will be appended to the supplied image name if the supplied name does
not specify and existing file.  As with paths, these extensions will
be searched in the order they are given.  Comments are any portion of
a line following a hash-mark (#).  The following is a sample
~/.xloadimagerc file:
.PP
.nf
  # paths to look for images in
  path= /usr/local/images
        /home/usr1/guest/madd/images
        /usr/include/X11/bitmaps

  # default extensions for images; .Z is automatic; scanned in order
  extension= .csun .msun .sun .face .xbm .bm
.fi
.SH IMAGE TYPES
.PP
\fIXloadimage\fR currently supports the following image types:
.nf
  Faces Project images
  Portable Bitmap (PBM) images
  Sun monochrome rasterfiles
  Sun color RGB rasterfiles
  X10 bitmap files
  X11 bitmap files
.fi
.PP
Both normal and compact PBM images are supported.  Both standard and
run-length encoded Sun rasterfiles are supported.
.SH AUTHOR
.nf
Jim Frost
Software Tool & Die
madd@std.com
.fi
.SH BUGS
Loading images onto the root with PseudoColor or GrayScale displays
can cause colormap problems (and may interfere with window manager
operation) if there are not enough colors in the default colormap to
allocate all of the colors read/write.  This can happen on images
which have too many unique colors or if images are loaded onto the
root in succession.  Since there is currently no X standard for
changing the root colormap, this problem may or may not be corrected
in the future.
.PP
Since \fIxloadimage\fR does not want a window larger than the image to
be displayed, it resizes the window to fit the image if the window
manager resizes the window to a size which is too large.  This could
cause a conflict if the window manager responds to the resize request
by resizing the window to the larger size.  This should be rare given
the state of current window managers and has never been observed.


contrib/clients/xloadimage/xloadimage.c
/* loadimage.c:
 *
 * generic image loader for X11
 *
 * jim frost 09.27.89
 *
 * Copyright 1989 Jim Frost.  See included file "copyright.h" for complete
 * copyright information.
 */

#include "copyright.h"
#include "xloadimage.h"

/* options array and definitions
 */

char *Options[] = {
  "onroot", /* global options */
  "border",
  "display",
  "geometry",
  "help",
  "identify",
  "list",
  "install",
  "path",
  "quiet",
  "supported",
  "verbose",
  "view",

  "at", /* image options */
  "background",
  "brighten",
  "center",
  "clip",
  "colors",
  "dither",
  "foreground",
  "name",
  "xzoom",
  "yzoom",
  "zoom",
  NULL
};

#define ONROOT    0
#define BORDER    1
#define DISPLAY   2
#define GEOMETRY  3
#define HELP      4
#define IDENTIFY  5
#define LIST      6
#define INSTALL   7
#define PATH      8
#define QUIET     9
#define SUPPORTED 10
#define VERBOSE   11
#define VIEW      12

#define AT         13
#define BACKGROUND 14
#define BRIGHT     15
#define CENTER     16
#define CLIP       17
#define COLORS     18
#define DITHER     19
#define FOREGROUND 20
#define NAME       21
#define XZOOM      22
#define YZOOM      23
#define ZOOM       24

/* the real thing
 */

main(argc, argv)
     int argc;
     char *argv[];
{ unsigned int  onroot= 1;
  char         *border;
  char         *dname;
  unsigned int  identify;
  unsigned int  install;
  unsigned int  verbose;
  Image        *dispimage;      /* image that will be sent to the display */
  Image        *newimage;       /* new image we're loading */
  Display      *disp;           /* display we're sending to */
  int           scrn;           /* screen we're sending to */
  XColor        xcolor;         /* color for border option */
  ImageOptions  images[MAXIMAGES]; /* list of image w/ options to load */
  unsigned int  imagecount;     /* number of images in ImageName array */
  unsigned int  a;
  unsigned int  winx, winy;     /* location of window */
  unsigned int  winwidth, winheight; /* geometry of window */

  if (argc < 2)
    usage(argv[0]);

  /* defaults and other initial settings.  some of these depend on what
   * our name was when invoked.
   */

  loadPathsAndExts();
  onroot= 0;
  verbose= 1;
  if (!strcmp(tail(argv[0]), "xview")) {
    onroot= 0;
    verbose= 1;
  }
  else if (!strcmp(tail(argv[0]), "xsetbg")) {
    onroot= 1;
    verbose= 0;
  }
  border= NULL;
  dname= NULL;
  identify= 0;
  install= 0;
  winx= winy= winwidth= winheight= 0;

  imagecount= 0;
  for (a= 1; a < argc; a++) {
    switch (optionNumber(argv[a], Options)) {
    case OPT_BADOPT:
      printf("%s: Bad option\n", argv[a]);
      usage(argv[0]);
      /* NOTREACHED */

    case OPT_NOTOPT:
      if (imagecount == MAXIMAGES)
	printf("%s: Too many images (ignoring)\n", argv[++a]);
      else
	images[imagecount++].name= argv[a];
      break;

    case OPT_SHORTOPT:
      printf("%s: Not enough characters to identify option\n", argv[a]);
      usage(argv[0]);
      /* NOTREACHED */

    /* process options global to everything
     */

    case ONROOT:
      onroot= 1;
      break;

    case BORDER:
      border= argv[++a];
      break;

    case DISPLAY:
      dname= argv[++a];
      break;

    case GEOMETRY:
      XParseGeometry(argv[++a], &winx, &winy, &winwidth, &winheight);
      break;

    case HELP:
      usage(argv[0]);
      exit(0);

    case IDENTIFY:
      identify= 1;
      break;

    case LIST:
      listImages();
      exit(0);

    case INSTALL:
      install= 1;
      break;

    case PATH:
      showPath();
      break;

    case QUIET:
      verbose= 0;
      break;

    case SUPPORTED:
      supportedImageTypes();
      break;

    case VERBOSE:
      verbose= 1;
      break;

    case VIEW:
      onroot= 0;

    /* process options local to an image
     */

    case AT:
      if (sscanf(argv[++a], "%d,%d",
		 &images[imagecount].atx, &images[imagecount].aty) != 2) {
	printf("Bad argument to -at\n");
	usage(argv[0]);
	/* NOTREACHED */
      }
      break;

    case BACKGROUND:
      images[imagecount].bg= argv[++a];
      break;

    case BRIGHT:
      images[imagecount].bright= atoi(argv[++a]);
      break;

    case CENTER:
      images[imagecount].center= 1;
      break;

    case CLIP:
      if (sscanf(argv[++a], "%d,%d,%d,%d",
		 &images[imagecount].clipx, &images[imagecount].clipy,
		 &images[imagecount].clipw, &images[imagecount].cliph) != 4) {
	printf("Bad argument to -clip\n");
	usage(argv[0]);
	/* NOTREACHED */
      }
      break;

    case COLORS:
      images[imagecount].colors= atoi(argv[++a]);
      if (images[imagecount].colors < 2) {
	printf("Argument to -colors is too low (ignored)\n");
	images[imagecount].colors= 0;
      }
      else if (images[imagecount].colors > 65536) {
	printf("Argument to -colors is too high (ignored)\n");
	images[imagecount].colors= 0;
      }
      break;

    case DITHER:
      images[imagecount].dither= 1;
      break;

    case FOREGROUND:
      images[imagecount].fg= argv[++a];
      break;

    case NAME:
      if (imagecount == MAXIMAGES)
	printf("%s: Too many images (ignoring)\n", argv[++a]);
      else
	images[imagecount++].name= argv[++a];
      break;

    case XZOOM:
      images[imagecount].xzoom= atoi(argv[++a]);
      break;

    case YZOOM:
      images[imagecount].yzoom= atoi(argv[++a]);
      break;

    case ZOOM:
      images[imagecount].xzoom= images[imagecount].yzoom= atoi(argv[++a]);
      break;

    default:

      /* this should not happen!
       */

      printf("%s: Internal error parsing arguments\n", argv[0]);
      exit(1);
    }
  }

  if (!imagecount) /* NO-OP from here on */
    exit(0);

  if (identify) {                    /* identify the named image(s) */
    for (a= 0; a < imagecount; a++)
      identifyImage(images[a].name);
    exit(0);
  }

  /* start talking to the display
   */

  if (! (disp= XOpenDisplay(dname))) {
    printf("%s: Cannot open display\n", XDisplayName(dname));
    exit(1);
  }
  scrn= DefaultScreen(disp);
  XSetIOErrorHandler(ioErrorHandler);

  dispimage= NULL;
  if (onroot && (winwidth || winheight || images[0].center ||
      images[a].atx || images[a].aty)) {
    if (!winwidth)
	winwidth= DisplayWidth(disp, scrn);
    if (!winheight)
      winheight= DisplayHeight(disp, scrn);
    if (DefaultDepth(disp, scrn) == 1)
      dispimage= newBitImage(winwidth, winheight);
    else {
      dispimage= newRGBImage(winwidth, winheight, DefaultDepth(disp, scrn));
      dispimage->rgb.used= 1;
    }
    *(dispimage->rgb.red)= 65535;   /* default border value is white */
    *(dispimage->rgb.green)= 65535;
    *(dispimage->rgb.blue)= 65535;
    if (border) {
      XParseColor(disp, DefaultColormap(disp, scrn), border, &xcolor);
      *dispimage->rgb.red= xcolor.red;
      *dispimage->rgb.green= xcolor.green;
      *dispimage->rgb.blue= xcolor.blue;
    }
    fill(dispimage, 0, 0, winwidth, winheight, 0, verbose);
  }

  /* load in each named image
   */

  for (a= 0; a < imagecount; a++) {
    if (! (newimage= loadImage(images[a].name, verbose)))
      continue;
    if ((dispimage && BITMAPP(dispimage)) || (DefaultDepth(disp, scrn) == 1))
      images[a].dither= 1;
    newimage= processImage(disp, scrn, newimage, &images[a], verbose);
    if (!images[a].clipw && !images[a].cliph) {
      images[a].clipw= newimage->width;
      images[a].cliph= newimage->height;
    }
    if (images[a].center) {
      images[a].atx= (dispimage->width - images[a].clipw) / 2;
      images[a].aty= (dispimage->height - images[a].cliph) / 2;
    }
    if (dispimage) {
      if (! dispimage->title)
	dispimage->title= dupString(newimage->title);
      merge(dispimage, newimage, images[a].atx, images[a].aty, verbose);
      freeImage(newimage);
    }
    else
      dispimage= newimage;
  }

  if (! dispimage) {
    printf("No images to display\n");
    exit(1);
  }

  if (onroot)
    imageOnRoot(disp, scrn, dispimage, install, verbose);
  else
    imageInWindow(disp, scrn, dispimage, winx, winy, winwidth, winheight,
		  install, verbose);
  XCloseDisplay(disp);
}
